source/simulator/stokes_matrix_free.cc

```diff
@@ -1420,69 +1465,111 @@ namespace aspect
+    mg::Matrix<vector_t > mg_interface_A(mg_interface_matrices_A);
+
+    // Schur complement matrix GMG
+    MGLevelObject<MatrixFreeOperators::MGInterfaceOperator<SchurComplementMatrixType> > mg_interface_matrices_Schur;
+    mg_interface_matrices_Schur.resize(0, sim.triangulation.n_global_levels()-1);
     for (unsigned int level=0; level<sim.triangulation.n_global_levels(); ++level)
-      mg_interface_matrices[level].initialize(mg_matrices[level]);
-    mg::Matrix<vector_t > mg_interface(mg_interface_matrices);
+      mg_interface_matrices_Schur[level].initialize(mg_matrices_Schur_complement[level]);
+    mg::Matrix<vector_t > mg_interface_Schur(mg_interface_matrices_Schur);
 
     // MG Matrix
-    mg::Matrix<vector_t > mg_matrix(mg_matrices);
+    mg::Matrix<vector_t > mg_matrix_A(mg_matrices_A_block);
+    mg::Matrix<vector_t > mg_matrix_Schur(mg_matrices_Schur_complement);
 
     // MG object
-    Multigrid<vector_t > mg(mg_matrix,
-                            mg_coarse,
-                            mg_transfer,
-                            mg_smoother,
-                            mg_smoother);
-    mg.set_edge_matrices(mg_interface, mg_interface);
+    // ABlock GMG
+    Multigrid<vector_t > mg_A(mg_matrix_A,
+                              mg_coarse_A,
+                              mg_transfer_A_block,
+                              mg_smoother_A,
+                              mg_smoother_A);
+    mg_A.set_edge_matrices(mg_interface_A, mg_interface_A);
+
+    // Schur complement matrix GMG
+    Multigrid<vector_t > mg_Schur(mg_matrix_Schur,
+                              mg_coarse_Schur,
+                              mg_transfer_Schur_complement,
+                              mg_smoother_Schur,
+                              mg_smoother_Schur);
+    mg_Schur.set_edge_matrices(mg_interface_Schur, mg_interface_Schur);
 
     // GMG Preconditioner
     typedef PreconditionMG<dim, vector_t, MGTransferMatrixFree<dim,double> > APreconditioner;
-    APreconditioner prec_A(dof_handler_v, mg, mg_transfer);
+    APreconditioner prec_A(dof_handler_v, mg_A, mg_transfer_A_block);
 
-    // For the Mass matrix Preconditioner we choose a Chebyshev smoother setup
-    // in a similar way to the coarse grid solver.
-    typedef PreconditionChebyshev<MassMatrixType,vector_t> MassPreconditioner;
-    MassPreconditioner prec_S;
-    typename MassPreconditioner::AdditionalData prec_S_data;
-    prec_S_data.smoothing_range = 1e-3;
-    prec_S_data.degree = numbers::invalid_unsigned_int;
-    prec_S_data.eig_cg_n_iterations = mass_matrix.m();
-    prec_S_data.preconditioner = mass_matrix.get_matrix_diagonal_inverse();
-    prec_S.initialize(mass_matrix,prec_S_data);
+    // Schur complement matrix GMG
+    typedef PreconditionMG<dim, vector_t, MGTransferMatrixFree<dim,double> > SchurPreconditioner;
+    APreconditioner prec_Schur(dof_handler_p, mg_Schur, mg_transfer_Schur_complement);
```

Comment:
```suggestion
    SchurPreconditioner prec_Schur(dof_handler_p, mg_Schur, mg_transfer_Schur_complement);
```

Reply:
This is a good catch. Now we only technically need one `typedef` since both preconditioners are of type `PreconditionMG<dim,dealii::LinearAlgebra::distributed::Vector<double>,MGTransferMatrixFree<dim,double>`. I just have `typedef PreconditionMG<...> GMGPreconditioner` for both now. Let me know if you like this fix, or still prefer to have the two different names (I can see a reason for both).